File: LearnGraphQL/gographQL/services/book_service.go
package services

import (
	"gographql/database"
	"gographql/models"
	"gographql/payload/request"
	"log"
)

type BookService interface {
	GetAllBooks() ([]*models.Book, error)
	GetBookByID(id string) (*models.Book, error)
	CreateBook(book *models.Book) error
	UpdateBook(book *models.Book) error
	DeleteBook(id string) error
	UpdateBooksPageCount(inputs []*request.BookPageCountInput) ([]*models.Book, error)
	SearchBooks(name, authorFirstName, authorLastName string, minPageCount, maxPageCount int) ([]*models.Book, error)
}

type BookServiceImpl struct{}

func (b BookServiceImpl) GetAllBooks() ([]*models.Book, error) {
	db, err := database.Connect()
	if err != nil {
		log.Fatal(err)
	}
	defer db.Close()

	rows, err := db.Query("SELECT id, name, page_count, author_id FROM book")
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var books []*models.Book
	for rows.Next() {
		var book models.Book
		err := rows.Scan(&book.ID, &book.Name, &book.PageCount, &book.Author.ID)
		if err != nil {
			return nil, err
		}
		books = append(books, &book)
	}
	return books, nil
}

func (b BookServiceImpl) GetBookByID(id string) (*models.Book, error) {
	db, err := database.Connect()
	if err != nil {
		log.Fatal(err)
	}
	defer db.Close()

	var book models.Book
	err = db.QueryRow("SELECT id, name, page_count, author_id FROM book WHERE id = ?", id).Scan(&book.ID, &book.Name, &book.PageCount, &book.Author.ID)
	if err != nil {
		return nil, err
	}
	return &book, nil
}

func (b BookServiceImpl) CreateBook(book *models.Book) error {
	db, err := database.Connect()
	if err != nil {
		log.Fatal(err)
	}
	defer db.Close()

	_, err = db.Exec("INSERT INTO book (id, name, page_count, author_id) VALUES (?, ?, ?, ?)", book.ID, book.Name, book.PageCount, book.Author.ID)
	if err != nil {
		return err
	}
	return nil
}

func (b BookServiceImpl) UpdateBook(book *models.Book) error {
	db, err := database.Connect()
	if err != nil {
		log.Fatal(err)
	}
	defer db.Close()

	_, err = db.Exec("UPDATE book SET name = ?, page_count = ?, author_id = ? WHERE id = ?", book.Name, book.PageCount, book.Author.ID, book.ID)
	if err != nil {
		return err
	}
	return nil
}

func (b BookServiceImpl) DeleteBook(id string) error {
	db, err := database.Connect()
	if err != nil {
		log.Fatal(err)
	}
	defer db.Close()

	_, err = db.Exec("DELETE FROM book WHERE id = ?", id)
	if err != nil {
		return err
	}
	return nil
}

func (b BookServiceImpl) UpdateBooksPageCount(inputs []*request.BookPageCountInput) ([]*models.Book, error) {
	//db, err := database.Connect()
	//if err != nil {
	//	log.Fatal(err)
	//}
	//defer db.Close()
	//
	//tx, err := db.Begin()
	//if err != nil {
	//	return nil, err
	//}
	//
	//for _, input := range inputs {
	//	_, err := tx.Exec("UPDATE book SET page_count = ? WHERE id = ?", input.PageCount, input.ID)
	//	if err != nil {
	//		tx.Rollback()
	//		return nil, err
	//	}
	//}
	//
	//err = tx.Commit()
	//if err != nil {
	//	return nil, err
	//}
	//
	//// Fetch updated books
	//var bookIDs []interface{}
	//for _, input := range inputs {
	//	bookIDs = append(bookIDs, input.ID)
	//}
	//
	//query, args, err := sqlx.In("SELECT id, name, page_count, author_id FROM book WHERE id IN (?)", bookIDs)
	//if err != nil {
	//	return nil, err
	//}
	//query = db.Rebind(query)
	//
	//rows, err := db.Query(query, args...)
	//if err != nil {
	//	return nil, err
	//}
	//defer rows.Close()
	//
	//var books []*models.Book
	//for rows.Next() {
	//	var book models.Book
	//	err := rows.Scan(&book.ID, &book.Name, &book.PageCount, &book.Author.ID)
	//	if err != nil {
	//		return nil, err
	//	}
	//	books = append(books, &book)
	//}
	return nil, nil
}

func (b BookServiceImpl) SearchBooks(name, authorFirstName, authorLastName string, minPageCount, maxPageCount int) ([]*models.Book, error) {
	// TODO: Implement search books functionality
	panic("implement me")
}

func NewBookService() BookService {
	return &BookServiceImpl{}
}

File: LearnGraphQL/springgraphQL/src/main/java/com/app/springgraphQL/service/BookService.java
package com.app.springgraphQL.service;

import com.app.springgraphQL.model.Author;
import com.app.springgraphQL.model.Book;
import com.app.springgraphQL.payload.request.BookPageCountInput;
import com.app.springgraphQL.repository.AuthorRepository;
import com.app.springgraphQL.repository.BookRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.stream.Collectors;

@Service
public class BookService {
	@Autowired
	private BookRepository bookRepository;
	@Autowired
	private AuthorRepository authorRepository;
	
	public List<Book> getAllBooks() {
		return bookRepository.findAll();
	}
	
	public Book getBookById(String id) {
		return bookRepository.findById(id).orElse(null);
	}
	
	public Book createBook(String name, int pageCount, String authorId) {
		Author author = authorRepository.findById(authorId).orElse(null);
		if(author == null) {
			throw new IllegalArgumentException("Author not found");
		}
		Book book = new Book(name, pageCount, author);
		return bookRepository.save(book);
	}
	
	public Book updateBook(String id, String name, int pageCount, String authorId) {
		Book book = bookRepository.findById(id).orElse(null);
		Author author = authorRepository.findById(authorId).orElse(null);
		if(book != null && author != null) {
			book.setName(name);
			book.setPageCount(pageCount);
			book.setAuthor(author);
			bookRepository.save(book);
		}
		return book;
	}
	/*TODO:
	*   chức năng phức tạp hơn thể hiện rõ ưu điểm của GraphQL so với REST, chúng ta sẽ tạo một truy vấn (query) và một mutation phức tạp hơn, bao gồm cả việc lồng các mối quan hệ và sử dụng các tính năng nâng cao của GraphQL như batch loading và data fetching optimization.*/
	//TODO:cho phép tìm kiếm sách dựa trên nhiều tiêu chí và hiển thị thông tin chi tiết về tác giả cùng với sách đó.
	public List<Book> searchBooks(String name, String authorFirstName, String authorLastName, Integer minPageCount, Integer maxPageCount) {
		// Logic tìm kiếm phức tạp sử dụng các tiêu chí đã cung cấp
		return bookRepository.findBooksByCriteria(name, authorFirstName, authorLastName, minPageCount, maxPageCount);
	}
    //TODO:  thêm một tính năng để cập nhật số lượng trang của nhiều sách cùng một lúc (batch update).
	 public List<Book> updateBooksPageCount(List<BookPageCountInput> booksPageCountInput) {
        List<String> bookIds = booksPageCountInput.stream()
                .map(BookPageCountInput::getId)
                .collect(Collectors.toList());

        List<Book> books = bookRepository.findAllById(bookIds);

        for (Book book : books) {
            booksPageCountInput.stream()
                    .filter(input -> input.getId().equals(book.getId()))
                    .findFirst()
                    .ifPresent(input -> book.setPageCount(input.getPageCount()));
        }

        return bookRepository.saveAll(books);
    }
	public void deleteBook(String id) {
		bookRepository.deleteById(id);
	}
}
